snapcraft/parts/plugins/kernel_build.sh

```diff
@@ -0,0 +1,323 @@
+    configs=*)     configs=${arg#*=}     ;; # valid: list, trumps kconfig-flavour
+    enable-zfs=*)  enable_zfs=${arg#*=}  ;; # valid: list, true|false
+    enable-perf=*) enable_perf=${arg#*=} ;; # valid: list, true|false
+  esac
+done
+
+# Set of valid targets specified by plugin
+build_target="$KERNEL_IMAGE $KERNEL_TARGET"
+
+# Remove artifacts from prior builds
+cleanup() {
+  printf 'Cleaning previous build first...\n'
+
+  rm -rf "${CRAFT_PART_INSTALL}/modules"
+  unlink "${CRAFT_PART_INSTALL}/lib/modules"
+}
+
+# Create specified defconfig
+gen_defconfig() {
+  make -j1                    \
+       -C "$CRAFT_PART_SRC"   \
+        O="$CRAFT_PART_BUILD" \
+        "$defconfig"
+}
+
+# Update config with flavour
+gen_flavour_config() {
+  OLDPWD="$PWD"
+  ubuntuconfig="${CRAFT_PART_SRC}/CONFIGS/${CRAFT_ARCH_BUILD_FOR}-config.flavour.${flavour}"
+  cd "${CRAFT_PART_SRC}"
+
+  # Generate the configs
+  # Don't let failures stop us
+  fakeroot debian/rules clean genconfigs || true
+
+  cat "$ubuntuconfig" > "${CRAFT_PART_BUILD}/.config"
+
+  fakeroot debian/rules clean
+
+  rm -rf CONFIGS
+
+  cd "$OLDPWD"
+}
+
+# Add any specified kconfig overrides
+add_kconfigs() {
+  kconfigs="$1"
+  _kconfigs="$(echo "$kconfigs" | sed -e 's/,/\n/g')"
+
+  printf 'Applying extra config....'
+
+  # Need to echo the list twice to force all Kconfigs to pickup the change
+  echo  "$_kconfigs"                    >  "${CRAFT_PART_BUILD}/.config_snap"
+  cat   "${CRAFT_PART_BUILD}/.config"   >> "${CRAFT_PART_BUILD}/.config_snap"
+  echo  "$_kconfigs"                    >> "${CRAFT_PART_BUILD}/.config_snap"
+  mv -f "${CRAFT_PART_BUILD}/.config_snap" "${CRAFT_PART_BUILD}/.config"
+}
+
+# Make sure the config is valid
+remake_config() {
+  bash -c 'yes "" || true' |
+    make -j1                    \
+         -C "$CRAFT_PART_SRC"   \
+          O="$CRAFT_PART_BUILD" \
+          oldconfig
+}
+
+# Check for Ubuntu Core and snap specific options
+check_config() {
+  echo "$_required_boot"     \
+       "$_required_generic"  \
+       "$_required_security" \
+       "$_required_snappy"   \
+       "$_required_systemd"  | while read -r config; do
+    if ! grep "^CONFIG_${config}=" "${CRAFT_PART_BUILD}/.config"; then
+      printf '*** WARNING ***\n'
+      printf 'Your kernel config is missing:\n'
+      printf '%s\n' "$config"
+      printf 'Which is recommended or required by Ubuntu Core!\n'
+      printf 'Continuing, but you should check your .config\n'
+    fi
+  done
+}
+
+# Gather kernel release info to encode in kernel artifacts
+release_info() {
+  printf 'Gathering release information\n'
+  DEBIAN="${CRAFT_PART_SRC}/debian"
+  src_pkg_name=$(sed -n '1s/^\(.*\) (.*).*$/\1/p'                                    "${DEBIAN}/changelog")
+  release=$(     sed -n '1s/^'"${src_pkg_name}"'.*(\(.*\)-.*).*$/\1/p'               "${DEBIAN}/changelog")
+  revision=$(    sed -n '1s/^'"${src_pkg_name}"'\ .*('"${release}"'-\(.*\)).*$/\1/p' "${DEBIAN}/changelog")
```

Comment:
This is all basically a cute way to grab the kernel version of the latest change -- for example, `6.8.0-60-63`. Format is... `<Package name> (kernel version) ...` So there's probably an easier way to grab this than some incredibly ungrokable `sed`s.